src/compiler/scala/tools/nsc/transform/CleanUp.scala

```diff
@@ -485,6 +485,11 @@ abstract class CleanUp extends Statics with Transform with ast.TreeDSL {
       case Apply(appMeth, elem0 :: Apply(wrapArrayMeth, (rest @ ArrayValue(elemtpt, _)) :: Nil) :: Nil)
       if wrapArrayMeth.symbol == wrapVarargsArrayMethod(elemtpt.tpe) && appMeth.symbol == ArrayModule_apply(elemtpt.tpe) =>
         treeCopy.ArrayValue(rest, rest.elemtpt, elem0 :: rest.elems).transform(this)
+      case Apply(appMeth, elem :: (nil: RefTree) :: Nil)
+      if nil.symbol == NilModule && appMeth.symbol == ArrayModule_apply(elem.tpe.widen) =>
```

Comment:
`&& treeInfo.isSafeToElide(nil)` to be sure there isn't some `RefTree` that also represents side effects.

I can't provoke the failure I'm worried about though.

```
scala> Array.apply[String]({???; scala.`package`}.Nil: _*)
scala.NotImplementedError: an implementation is missing
  at scala.Predef$.$qmark$qmark$qmark(Predef.scala:347)
  ... 28 elided

```

Reply:
That's not going to work because `q"scala.package.Nil".symbol` isn't `NilModule`, it's the symbol of that val. I don;'t think it's possible to hit this (a `RefTree` that has symbol `NilModule` at this point can't be a `Select(something.side.effecty, "Nil")`) because there's no way to return `scala.collection.immutable` from a block as you've got. I'll change it out of an abundance of caution anyways.

Reply:
This is what I was trying to find:

```
scala> def foo = {???; Array}.apply[AnyRef](Nil)
def foo: Array[AnyRef]

scala> foo
val res0: Array[AnyRef] = Array(List())
```

A ` && treeInfo.isQualifierSafeToElide(appMeth)` should do the trick.

Reply:
I've pushed this suggestion to the new and existing cases.